RecoTracker/FinalTrackSelectors/plugins/TrackCutClassifier.cc

```diff
@@ -48,61 +48,161 @@ namespace {
+      fillArrayF(dz_par1,      dz_par,"dz_par1");
+      fillArrayF(dz_par2,      dz_par,"dz_par2");
+      edm::ParameterSet dr_par = cfg.getParameter<edm::ParameterSet>("dr_par");
+      fillArrayF(dr_par1,      dr_par,"dr_par1");
+      fillArrayF(dr_par2,      dr_par,"dr_par2");
+      fillArrayF(d0err,        dr_par,"d0err");
+      fillArrayF(d0err_par,    dr_par,"d0err_par");
     }
     
     
     
     float operator()(reco::Track const & trk,
-		   reco::BeamSpot const & beamSpot,
-		   reco::VertexCollection const & vertices,
-		   GBRForest const *) const {
-
-     
+		     reco::BeamSpot const & beamSpot,
+		     reco::VertexCollection const & vertices,
+		     GBRForest const *) const {
       
       float ret = 1.f;
+      float dummy[3] = {1E-5, 1E-5, 1E-5};
+      ret = std::min(ret,cut(float(trk.ndof()),dummy,std::greater_equal<float>()) );
+
       auto  nLayers = trk.hitPattern().trackerLayersWithMeasurement();
       ret = std::min(ret,cut(nLayers,minLayers,std::greater_equal<int>()));
       if (ret==-1.f) return ret;
 
-      ret = std::min(ret,cut(chi2(trk)/float(nLayers),maxChi2n,std::less_equal<float>()));
+      ret = std::min(ret,cut(chi2n(trk)/float(nLayers),maxChi2n,std::less_equal<float>()));
       if (ret==-1.f) return ret;
-      ret = std::min(ret,cut(chi2(trk),maxChi2,std::less_equal<float>()));
+
+      ret = std::min(ret,cut(chi2n(trk),maxChi2,std::less_equal<float>()));
       if (ret==-1.f) return ret;
      
       ret = std::min(ret,cut(n3DLayers(trk),min3DLayers,std::greater_equal<int>()));
       if (ret==-1.f) return ret;
+
       ret = std::min(ret,cut(nPixelHits(trk),minPixelHits,std::greater_equal<int>()));
       if (ret==-1.f) return ret;
+
       ret = std::min(ret,cut(lostLayers(trk),maxLostLayers,std::less_equal<int>()));
-     
+      if (ret==-1.f) return ret;
+      
+      // original dz and dr cut
       if (maxDz[2]<std::numeric_limits<float>::max() || maxDr[2]<std::numeric_limits<float>::max()) {
-        if (ret==-1.f) return ret;
-        Point bestVertex = getBestVertex(trk,vertices);
-        ret = std::min(ret,cut(dz(trk,bestVertex), maxDz,std::less_equal<float>()));
-        ret = std::min(ret,cut(dr(trk,bestVertex), maxDr,std::less_equal<float>()));
+
+	// if not primaryVertices are reconstructed, check compatibility w.r.t. beam spot
+        Point bestVertex = getBestVertex(trk,vertices,minNVtxTrk); // min number of tracks [2 (=default) for offline, 3 for HLT]
+	float maxDzcut[3];
+	std::copy(std::begin(maxDz),std::end(maxDz),std::begin(maxDzcut));
+	if (bestVertex.z() < -99998.) {
+	  bestVertex = beamSpot.position();
+	  std::copy(std::begin(maxDzWrtBS),std::end(maxDzWrtBS),std::begin(maxDzcut));
+	}
+	ret = std::min(ret,cut(dr(trk,bestVertex), maxDr,std::less<float>()));
+	if (ret==-1.f) return ret;
+
+	ret = std::min(ret,cut(dz(trk,bestVertex), maxDzcut,std::less<float>()));
+	if (ret==-1.f) return ret;
```

Comment:
https://github.com/mtosi/cmssw/blob/from-CMSSW_8_0_0_pre4_updateTrackCutSelector/RecoTracker/FinalTrackSelectors/plugins/TrackCutClassifier.cc#L27-L35

you can get a result of 
- -1, which means an AND of all failed cut for index 0. This is independent of the order.
-  ```{-.5, .5, 1.} ``` for one of the three passing cuts applied on the *last* variable.

Any value of ret different from -1 is defined by the last applied cut.